Don't create fragmented blob made of other blobs with fetch data

jyavenard opened the pull request:
#### fcddb4ea71423066c6f2d9a4ab4bd0b268a962df
<pre>
Don&apos;t create fragmented blob made of other blobs with fetch data
<a href="https://bugs.webkit.org/show_bug.cgi?id=296531">https://bugs.webkit.org/show_bug.cgi?id=296531</a>
<a href="https://rdar.apple.com/156827880">rdar://156827880</a>

Reviewed by Youenn Fablet.

Follow-up on 297885@main. Instead of constructing a Blob
made of multiple blobs, which would create a lot of Blob URLs, we instead
add the ability to create a Blob from a FragmentedSharedBuffer. A BlobPart
is also amended to allow for holding a SharedBuffer.

No change in JS observable behaviours. Covered by existing tests.

* Source/WebCore/platform/network/BlobPart.h:
(WebCore::BlobPart::BlobPart): Add SharedBuffer as variant, clean-up
code to follow SaferCPP coding guidelines.
(WebCore::BlobPart::takeData): Renamed for consistency and better indicate how the code was used.
(WebCore::BlobPart::moveData): Deleted.

Canonical link: <a href="https://commits.webkit.org/297920@main">https://commits.webkit.org/297920@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/15717b5cff5d45805a523d095a6c22e33dab90fd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113468 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33184 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23633 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119652 "Built successfully") | [💥 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64239 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115377 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33803 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41756 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86335 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/64239 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116415 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26997 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102017 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66663 "Found 144 new API test failures: /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/configuration, /WPE/TestBackForwardList:/webkit/WebKitWebView/navigation-after-session-restore, /WPE/TestInputMethodContext:/webkit/WebKitInputMethodContext/reset, /WPE/TestInputMethodContext:/webkit/WebKitInputMethodContext/surrounding, /WPE/TestLoaderClient:/webkit/WebKitURIResponse/http-headers, /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/timezone, /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/timezone-worker, /WPE/TestWebKitUserContentManager:/webkit/WebKitUserContentManager/injected-script, /WPE/TestLoaderClient:/webkit/WebKitWebView/unfinished-subresource-load, /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/cookies ... (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26264 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20137 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63385 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96380 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20215 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122888 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40494 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30250 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95196 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40885 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98221 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94942 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24215 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40064 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17878 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/36661 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40379 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45879 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40036 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43351 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41789 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/fileapi/BlobBuilder.cpp:
```diff
@@ -84,6 +84,16 @@ void BlobBuilder::append(const String& text)
+void BlobBuilder::append(Ref<FragmentedSharedBuffer>&& buffer)
+{
+    if (!m_appendableData.isEmpty())
+        m_items.append(BlobPart(WTFMove(m_appendableData)));
```
Would be better to std::exchange m_appendableData.

youennf commented on Source/WebCore/platform/network/BlobPart.h:
```diff
@@ -37,6 +38,8 @@ class BlobPart {
 private:
     friend struct IPC::ArgumentCoder<BlobPart, void>;
 public:
+    using VariantType = Variant<Vector<uint8_t>, RefPtr<SharedBuffer>, URL>;
```
Can we use Ref<SharedBuffer>?

jyavenard replied:
Whatever earlier version gave me an error with the default constructor. But somehow now it's okay. so I made it a Ref again

youennf commented on Source/WebCore/platform/network/BlobPart.h:
```diff
@@ -81,12 +92,12 @@ class BlobPart {
 
 private:
-    BlobPart(Variant<Vector<uint8_t>, URL>&& dataOrURL)
+    BlobPart(VariantType&& dataOrURL)
```
explicit

youennf commented on Source/WebCore/platform/network/BlobPart.h:
```diff
@@ -62,10 +70,13 @@ class BlobPart {
+        if (std::holds_alternative<Vector<uint8_t>>(m_dataOrURL))
+            return std::get<Vector<uint8_t>>(std::exchange(m_dataOrURL, Vector<uint8_t> { }));
+        RefPtr buffer = std::get<RefPtr<SharedBuffer>>(std::exchange(m_dataOrURL, Vector<uint8_t> { }));
+        return buffer->extractData();
```
I would tend to use WTF::switchOn.

jyavenard committed:
Don't create fragmented blob made of other blobs with fetch data
https://bugs.webkit.org/show_bug.cgi?id=296531
rdar://156827880

Reviewed by Youenn Fablet.

Follow-up on 297885@main. Instead of constructing a Blob
made of multiple blobs, which would create a lot of Blob URLs, we instead
add the ability to create a Blob from a FragmentedSharedBuffer. A BlobPart
is also amended to allow for holding a SharedBuffer.

No change in JS observable behaviours. Covered by existing tests.

* Source/WebCore/platform/network/BlobPart.h:
(WebCore::BlobPart::BlobPart): Add SharedBuffer as variant, clean-up
code to follow SaferCPP coding guidelines.
(WebCore::BlobPart::takeData): Renamed for consistency and better indicate how the code was used.
(WebCore::BlobPart::moveData): Deleted.

Canonical link: https://commits.webkit.org/297920@main
